torch/csrc/distributed/c10d/NCCLUtils.cpp

```diff
@@ -80,14 +64,30 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
   auto comm = std::make_shared<NCCLComm>();
   // This call will block until the source communicator is initialized
   auto sourceComm = source->getNcclComm();
+#ifndef NCCL_HAS_COMM_NONBLOCKING
   C10D_NCCL_CHECK(
       ncclCommSplit(sourceComm, color_id, rank, &(comm->ncclComm_), &config),
       std::nullopt);
+#else
+  // After calling ncclCommSplit in non-blocking mode, we should wait for the
```

Comment:
i'm kinda confused about the meaning of non-blocking mode in nccl.  maybe it would be a good time to write up a note on different modes of nccl init and what their semantics are.

are there 2 states that we transition through in non-blocking init?  1) ncclInProgress state, 2) still doing some more non-blocking init, finally 3) ready for use?  And the point is we need to block/wait on (1) but we can let (2) be asynchronous?  otherwise i'm confused about what part of non-blocking init is actually non-blocking

Reply:
> the meaning of non-blocking mode in nccl

In general, it means the API call will return immediately, instead of blocking the calling thread until the init, finalize, etc operations complete.

> are there 2 states that we transition through in non-blocking init? 1) ncclInProgress state, 2) still doing some more non-blocking init, finally 3) ready for use? 

In general, two states only: ncclInProgress and ncclSuccess (if there is no error).

`ncclCommSplit` is a special case: it involves two communicators -- the parent comm and the child comm.
As of NCCL 2.23 (today): when `ncclCommSplit` returns, a valid pointer for child comm may not have been assigned. We'd need to wait for `ncclSuccess` on the parent comm to confirm that. But that doesn't mean the child comm is ready for use, we need to wait for `ncclSuccess` on the child comm to confirm that.

I have talked to the NCCL team re improvement of the expectation, i.e. when `ncclCommSplit` returns, a valid pointer for child comm SHOULD have been assigned. It seems NCCL team agrees, so that the API presents non-blocking semantics for both the parent and the child comms. Detailed discussion is here:
https://github.com/NVIDIA/nccl/issues/1472